### PR TITLE
node:http.request should emit events in the right order

### PR DESCRIPTION
**Pull request by nektro**
Fixes https://github.com/oven-sh/bun/issues/4357
Fixes https://github.com/oven-sh/bun/issues/5470
Fixes https://github.com/oven-sh/bun/issues/7716

### CODE REVIEW

File: test/js/node/http/node-http.test.ts [hunk 1]
```diff
@@ -1804,4 +1806,33 @@ it("#10177 response.write with non-ascii latin1 should not cause duplicated char
       }
       finish();
     });
+}, 20_000);
+
+it("should emit events in the right order", async () => {
+  const { stdout, stderr, exited } = Bun.spawn({
+    cmd: [bunExe(), "run", path.join(import.meta.dir, "fixtures/log-events.mjs")],
+    stdout: "pipe",
+    stdin: "pipe",
```
Comment (Jarred-Sumner) [inline test/js/node/http/node-http.test.ts]:
```suggestion
    stdin: "ignore",
```

File: test/js/node/http/node-http.test.ts [hunk 2]
```diff
@@ -1804,4 +1806,33 @@ it("#10177 response.write with non-ascii latin1 should not cause duplicated char
...
+    stderr: "pipe",
+    env: bunEnv,
+  });
+  expect(stderr).toBeDefined();
```
Comment (Jarred-Sumner) [inline test/js/node/http/node-http.test.ts]: we shouldn't need to check that stderr/stdout is defined, that would be a bug in Bun.spawn if it weren't

```suggestion
  expect(await exited).toBe(0);
```

File: test/js/node/http/node-http.test.ts [hunk 3]
```diff
@@ -1804,4 +1806,33 @@ it("#10177 response.write with non-ascii latin1 should not cause duplicated char
...
+  const err = await new Response(stderr).text();
+  expect(err).toBeEmpty();
+  expect(stdout).toBeDefined();
```
Comment (Jarred-Sumner) [inline test/js/node/http/node-http.test.ts]:
```suggestion
```

File: src/js/node/http.ts [hunk 4]
```diff
@@ -1268,7 +1269,7 @@ ServerResponse.prototype.assignSocket = function (socket) {
     throw ERR_HTTP_SOCKET_ASSIGNED();
   }
   socket._httpMessage = this;
-  socket.on("close", onServerResponseClose);
+  socket.on("close", () => onServerResponseClose.$call(socket));
```
Comment (Jarred-Sumner) [inline src/js/node/http.ts:1271]: theres never an error event or anything passed along as arguments to socket.on("close"), right?

Comment (nektro) [reply src/js/node/http.ts:1271]: correct

**Comment by Jarred-Sumner** [comment]
The code looks fine, but there are several failing tests

File: src/js/node/stream.ts [hunk 5]
```diff
@@ -1890,7 +1888,7 @@ var require_destroy = __commonJS({
           r.errored = err;
         }
         if (sync) {
-          runOnNextTick(emitErrorNT, stream, err);
```
Comment (Jarred-Sumner) [inline src/js/node/stream.ts:1893]: This adds an extra property lookup in a very hot code path

**Comment by davidstevens37** [comment]
@nektro @Jarred-Sumner  Thank you for your hard work and contributions on this 🙏 !  It appears this has been included since v1.1.5 (please correct me if i'm wrong). 

I've just retested #7716 in both v.1.1.6 and v1.1.6-canary. However, the behavior presents as it does in the issue: the IncomingRequest events are not emitted when the client disconnects.

Is there something I might be misunderstanding about the expected behavior? If not, we might need to re-open #7716 as it's still prone to memory leaks caused by the server being unaware when a client connection is closed.

Thanks again and apologies if I've misunderstood something.

**Comment by nektro** [comment]
apologies, will take a look and confirm first thing in the morning

**Pull request merged**